File: Data Structures & Algorithms/replace-elements-with-greatest-element-on-right-side/submission-0.py
class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        n = len(arr)
        last_digit = arr[n-1]
        result = [0] * n
        result[n-1] = -1

        for i in range(n-2 , -1 , -1):
            result[i] =  last_digit
            last_digit = max(last_digit , arr[i])
        

        return result

            

File: Data Structures & Algorithms/eating-bananas/submission-0.go
func minEatingSpeed(piles []int, h int) int {
    left := 1
    right := maxArr(piles)
    ans := right

    for left <= right {
        mid := (left + right) / 2

        if canEat(piles, h, mid) {
            ans = mid
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return ans
}

func canEat(piles []int, h int, k int) bool {
    hours := 0
    for _, p := range piles {
        hours += (p + k - 1) / k
    }
    return hours <= h
} 

func maxArr(arr []int) int {
    max := arr[0]
    for i := 1; i < len(arr); i++ {
        if arr[i] > max {
            max = arr[i]
        }
    }
    return max
}